File: dice.pl
#!/usr/bin/perl

use strict;
use warnings;

use List::AllUtils;

print main();


sub main {
    my $args = parse(\@ARGV);

    my @polys;

    my $combinations = 1;

    for my $r (@{ $args->{rolls} }) {
	$combinations *= $r->{die} ** $r->{num};
	push(@polys, @{ mkroll($r) });
    }

    my $bigpoly = megamul(\@polys);
    my @final	= map { $_ / $combinations } @$bigpoly;

    my $min = List::AllUtils::firstidx { defined($_) && $_ > 0 } @final;
    my $max = $#final;

    my $output = "\n";

    for my $exp ($min .. $max) {
	$output .= sprintf(
	    "%-10d%-10.5f%-48s\n",
	    $exp + $args->{const},
	    $final[$exp],
	    "#" x sprintf("%.0f", 500 * $final[$exp])
	);
    }

    return $output . "\n";
}


sub parse ($) {
    my ($args) = @_;

    my $constant = 0;
    my @rolls;

    for my $t (@$args) {
	if ($t =~ /d/) {
	    my ($n, $d) = split('d', $t);
	    push(@rolls, { num => $n, die => $d });
	}
	else {
	    $constant += $t;
	}
    }

    return {
	const => $constant,
	rolls => \@rolls
    };
}


sub mkroll ($) {
    my ($r) = @_;

    my @results;

    for my $i (1 .. $r->{num}) {
	my @result;

	for my $j (1 .. $r->{die}) {
	    $result[$j] = 1;
	}

	push @results, \@result;
    }

    return \@results;
}


sub poly_multiply ($$) {
    my ($p1, $p2) = @_;

    my @result;

    for (my $i = 0; $i < @$p1; $i++) {
	for (my $j = 0; $j < @$p2; $j++) {
	    my $t1 = $p1->[$i] // 0;
	    my $t2 = $p2->[$j] // 0;
	    $result[ $i + $j ] += $t1 * $t2;
	}
    }

    return \@result;
}


sub megamul {
    my ($polys) = @_;

    my $result = shift(@$polys);

    for my $p (@$polys) {
	$result = poly_multiply($result, $p);
    }

    return $result;
}


1;

File: dice.js


var utils = require('util');

console.log( main() );


function cool_round (x, n) {
    var scalar = Math.pow(10, n);
    return Math.round( scalar * x ) / scalar;
}


function mkhg (x) {
    var num = Math.round(500 * x),
	result = []
    ;

    for (var i = 0; i < num; i++) {
	result.push("#");
    }

    return result.join("");
}


function main () {
    var args = parse(process.argv.splice(2)),
	polys = [],
	combinations = 1,
	bigpoly = [],
	min = 0,
	max = 0,
	output = "\n"
    ;

    args.rolls.forEach(
	function (e, i) {
	    min += e.num;
	    max += e.num * e.die;
	    combinations *= Math.pow(e.die, e.num);
	    polys = polys.concat( mkroll(e) );
	}
    );

    bigpoly = megamul(polys).map(
	function (x) {
	    return cool_round(x / combinations, 5);
	}
    );

    for (var exp = min; exp <= max; exp++) {
	output += utils.format(
	    "%d\t\t%d\t\t%s\n",
	    exp + args.const,
	    bigpoly[exp],
	    mkhg(bigpoly[exp])
	);
    }

    return output + "\n";
}


function parse (args) {
    var constant = 0;
	rolls = []
    ;

    args.forEach(
	function (arg) {
	    if (arg.match(/d/)) {
		var yarr = arg.split('d');
		    n	 = parseInt(yarr[0]),
		    d	 = parseInt(yarr[1])
		;

		rolls.push({ num : n, die : d });
	    }
	    else {
		constant += parseInt(arg);
	    }
	}
    );

    return {
	const : constant,
	rolls : rolls
    };
}


function mkroll (r) {
    var results = [];

    for (var i = 1; i <= r.num; i++) {
	var result = [];

	for (var j = 1; j <= r.die; j++) {
	    result[j] = 1;
	}

	results.push( result );
    }

    return results;
}


function poly_multiply (p1, p2) {
    var result = [];

    for (var i = 1; i < p1.length; i++) {
	for (var j = 1; j < p2.length; j++) {
	    var k = i + j;
	    var a = (p1[i] || 0) * (p2[j] || 0);
	    result[k]  = result[k] || 0;
	    result[k] += a;
	}
    }

    return result;
}


function megamul (polys) {
    var result = polys.shift();

    polys.forEach(
	function (p) {
	    result = poly_multiply(result, p);
	}
    );

    return result;
}


